File: src/scpn_control/control/multi_shot_campaign.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Control — Multi-shot pulsed-campaign orchestration.
"""Multi-shot campaign admission over pulsed scheduler and bank telemetry."""

from __future__ import annotations

import hashlib
import json
import math
import uuid
from dataclasses import dataclass
from typing import Any

from scpn_control.control.capacitor_bank_state import CapacitorBank, CapacitorBankSpec
from scpn_control.control.pulsed_scenario_scheduler_v2 import (
    CapacitorBankTelemetry,
    PulsedPlasmaTelemetry,
    PulsedScenarioCommand,
    PulsedScenarioScheduler,
    PulsedScenarioSpec,
)

MULTI_SHOT_CAMPAIGN_SCHEMA_VERSION = "scpn-control.multi-shot-campaign.v1"
_EXPECTED_TRANSITION_STATES = (
    "ramp_up",
    "flat_top",
    "burn",
    "expansion",
    "dump",
    "recharge",
    "cool_down",
    "idle",
)


@dataclass(frozen=True)
class CampaignShotSample:
    """One timestamped plasma and bank telemetry sample for a pulsed shot."""

    t_s: float
    plasma: PulsedPlasmaTelemetry
    bank: CapacitorBankTelemetry | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "t_s", _non_negative("t_s", self.t_s))
        if not isinstance(self.plasma, PulsedPlasmaTelemetry):
            raise TypeError("plasma must be PulsedPlasmaTelemetry")
        if self.bank is not None and not isinstance(self.bank, CapacitorBankTelemetry):
            raise TypeError("bank must be CapacitorBankTelemetry or None")


@dataclass(frozen=True)
class CampaignShotPlan:
    """Input contract for one shot in a multi-shot campaign."""

    shot_id: str
    samples: tuple[CampaignShotSample, ...]
    initial_bank_voltage_V: float
    initial_bank_current_A: float = 0.0

    def __post_init__(self) -> None:
        shot_id = self.shot_id.strip()
        if not shot_id:
            raise ValueError("shot_id must not be empty")
        if not self.samples:
            raise ValueError("samples must not be empty")
        object.__setattr__(self, "shot_id", shot_id)
        object.__setattr__(self, "samples", tuple(self.samples))
        object.__setattr__(
            self, "initial_bank_voltage_V", _non_negative("initial_bank_voltage_V", self.initial_bank_voltage_V)
        )
        object.__setattr__(
            self, "initial_bank_current_A", _finite("initial_bank_current_A", self.initial_bank_current_A)
        )


@dataclass(frozen=True)
class CampaignCommandLog:
    """Stable command log row emitted by the scheduler."""

    t_s: float
    state: str
    action: str
    reason: str
    transition: bool
    dwell_s: float

    @classmethod
    def from_command(cls, command: PulsedScenarioCommand) -> CampaignCommandLog:
        """Convert a scheduler command to a JSON-stable row."""
        return cls(
            t_s=command.t_s,
            state=command.state.value,
            action=command.action.value,
            reason=command.reason,
            transition=command.transition,
            dwell_s=command.dwell_s,
        )

    def to_json(self) -> dict[str, float | str | bool]:
        """Return the JSON representation."""
        return {
            "t_s": self.t_s,
            "state": self.state,
            "action": self.action,
            "reason": self.reason,
            "transition": self.transition,
            "dwell_s": self.dwell_s,
        }


@dataclass(frozen=True)
class CampaignShotResult:
    """Admission result for one pulsed shot."""

    shot_id: str
    pulse_id: str
    passed: bool
    failure_reason: str | None
    terminal_state: str
    transition_states: tuple[str, ...]
    command_log: tuple[CampaignCommandLog, ...]
    shot_phase_log: tuple[dict[str, float | str], ...]
    capacitor_state_initial_J: float
    capacitor_state_final_J: float
    energy_recovered_J: float
    trigger_timestamp_ns: int | None

    def to_json(self) -> dict[str, Any]:
        """Return the JSON representation."""
        return {
            "shot_id": self.shot_id,
            "pulse_id": self.pulse_id,
            "passed": self.passed,
            "failure_reason": self.failure_reason,
            "terminal_state": self.terminal_state,
            "transition_states": list(self.transition_states),
            "command_log": [row.to_json() for row in self.command_log],
            "shot_phase_log": list(self.shot_phase_log),
            "capacitor_state_initial_J": self.capacitor_state_initial_J,
            "capacitor_state_final_J": self.capacitor_state_final_J,
            "energy_recovered_J": self.energy_recovered_J,
            "trigger_timestamp_ns": self.trigger_timestamp_ns,
        }

    def replay_v1_1_extension(self) -> dict[str, Any]:
        """Return fields compatible with geometry-neutral replay v1.1 extensions."""
        return {
            "pulse_id": self.pulse_id,
            "capacitor_state_initial_J": self.capacitor_state_initial_J,
            "trigger_timestamp_ns": self.trigger_timestamp_ns,
            "energy_recovered_J": self.energy_recovered_J,
            "shot_phase_log": list(self.shot_phase_log),
        }


class MultiShotCampaignOrchestrator:
    """Run deterministic multi-shot admission over scheduler and bank contracts."""

    def __init__(
        self,
        campaign_id: str,
        scheduler_spec: PulsedScenarioSpec,
        bank_spec: CapacitorBankSpec,
        *,
        require_complete_lifecycle: bool = True,
    ) -> None:
        campaign = campaign_id.strip()
        if not campaign:
            raise ValueError("campaign_id must not be empty")
        if not isinstance(scheduler_spec, PulsedScenarioSpec):
            raise TypeError("scheduler_spec must be PulsedScenarioSpec")
        if not isinstance(bank_spec, CapacitorBankSpec):
            raise TypeError("bank_spec must be CapacitorBankSpec")
        self.campaign_id = campaign
        self.scheduler_spec = scheduler_spec
        self.bank_spec = bank_spec
        self.require_complete_lifecycle = bool(require_complete_lifecycle)

    def run(self, shots: tuple[CampaignShotPlan, ...] | list[CampaignShotPlan]) -> dict[str, Any]:
        """Run all shots and return a digest-bound campaign report."""
        shot_plans = tuple(shots)
        if not shot_plans:
            raise ValueError("shots must not be empty")
        seen: set[str] = set()
        results: list[CampaignShotResult] = []
        for shot in shot_plans:
            if not isinstance(shot, CampaignShotPlan):
                raise TypeError("shots must contain CampaignShotPlan entries")
            if shot.shot_id in seen:
                raise ValueError(f"duplicate shot_id: {shot.shot_id}")
            seen.add(shot.shot_id)
            results.append(self._run_shot(shot))
        passed_count = sum(1 for result in results if result.passed)
        report: dict[str, Any] = {
            "schema_version": MULTI_SHOT_CAMPAIGN_SCHEMA_VERSION,
            "campaign_id": self.campaign_id,
            "shot_count": len(results),
            "passed_count": passed_count,
            "failed_count": len(results) - passed_count,
            "require_complete_lifecycle": self.require_complete_lifecycle,
            "expected_transition_states": list(_EXPECTED_TRANSITION_STATES),
            "shots": [result.to_json() for result in results],
            "payload_sha256": "",
        }
        report["payload_sha256"] = _payload_sha256(report)
        return report

    def _run_shot(self, shot: CampaignShotPlan) -> CampaignShotResult:
        scheduler = PulsedScenarioScheduler(self.scheduler_spec)
        bank = CapacitorBank(
            self.bank_spec,
            initial_voltage_V=shot.initial_bank_voltage_V,
            initial_current_A=shot.initial_bank_current_A,
        )
        initial_energy = bank.state.energy_J
        last_energy = initial_energy
        min_energy = initial_energy
        command_log: list[CampaignCommandLog] = []
        failure_reason: str | None = None
        for sample in shot.samples:
            try:
                bank_telemetry = sample.bank if sample.bank is not None else bank.telemetry()
                last_energy = bank_telemetry.energy_J
                min_energy = min(min_energy, last_energy)
                command = scheduler.step(sample.t_s, sample.plasma, bank_telemetry)
            except ValueError as exc:
                failure_reason = str(exc)
                break
            command_log.append(CampaignCommandLog.from_command(command))

        phase_log = tuple(
            {
                "t": record.t_s,
                "state": record.to_state.value,
                "reason": record.reason,
            }
            for record in scheduler.audit_log
        )
        transition_states = tuple(str(row["state"]) for row in phase_log)
        terminal_state = scheduler.state.value
        if failure_reason is None:
            failure_reason = self._admission_failure_reason(terminal_state, transition_states)
        trigger_timestamp_ns = next(
            (int(round(float(row["t"]) * 1_000_000_000)) for row in phase_log if row["state"] == "burn"),
            None,
        )
        return CampaignShotResult(
            shot_id=shot.shot_id,
            pulse_id=_pulse_id(self.campaign_id, shot.shot_id),
            passed=failure_reason is None,
            failure_reason=failure_reason,
            terminal_state=terminal_state,
            transition_states=transition_states,
            command_log=tuple(command_log),
            shot_phase_log=phase_log,
            capacitor_state_initial_J=initial_energy,
            capacitor_state_final_J=last_energy,
            energy_recovered_J=max(last_energy - min_energy, 0.0),
            trigger_timestamp_ns=trigger_timestamp_ns,
        )

    def _admission_failure_reason(self, terminal_state: str, transition_states: tuple[str, ...]) -> str | None:
        if self.require_complete_lifecycle and transition_states != _EXPECTED_TRANSITION_STATES:
            return "shot did not traverse the complete pulsed lifecycle"
        if terminal_state != "idle":
            return "shot did not return to idle"
        return None


def _pulse_id(campaign_id: str, shot_id: str) -> str:
    return str(uuid.uuid5(uuid.NAMESPACE_URL, f"scpn-control:{campaign_id}:{shot_id}"))


def _stable_json(payload: Any) -> str:
    return json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=True)


def _payload_sha256(payload: dict[str, Any]) -> str:
    unsigned = dict(payload)
    unsigned["payload_sha256"] = ""
    return hashlib.sha256(_stable_json(unsigned).encode("utf-8")).hexdigest()


def _finite(name: str, value: float) -> float:
    number = float(value)
    if not math.isfinite(number):
        raise ValueError(f"{name} must be finite")
    return number


def _non_negative(name: str, value: float) -> float:
    number = _finite(name, value)
    if number < 0.0:
        raise ValueError(f"{name} must be non-negative")
    return number


__all__ = [
    "CampaignCommandLog",
    "CampaignShotPlan",
    "CampaignShotResult",
    "CampaignShotSample",
    "MULTI_SHOT_CAMPAIGN_SCHEMA_VERSION",
    "MultiShotCampaignOrchestrator",
]

File: scpn-control-rs/crates/control-control/src/multi_shot_campaign.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Control — Rust multi-shot pulsed-campaign orchestration.
//! Multi-shot campaign admission over pulsed scheduler and bank telemetry.

use crate::capacitor_bank::{CapacitorBank, CapacitorBankSpec};
use crate::pulsed_scenario::{
    CapacitorBankTelemetry, PulsedPlasmaTelemetry, PulsedScenarioCommand, PulsedScenarioError,
    PulsedScenarioScheduler, PulsedScenarioSpec,
};
use sha2::{Digest, Sha256};
use std::collections::HashSet;
use std::error::Error;
use std::fmt::{Display, Formatter};

const MULTI_SHOT_CAMPAIGN_SCHEMA_VERSION: &str = "scpn-control.multi-shot-campaign.v1";
const PULSED_MPC_DECISION_EVIDENCE_SCHEMA_VERSION: &str =
    "scpn-control.pulsed-mpc-decision-evidence.v1";
const EXPECTED_TRANSITION_STATES: [&str; 8] = [
    "ramp_up",
    "flat_top",
    "burn",
    "expansion",
    "dump",
    "recharge",
    "cool_down",
    "idle",
];

/// One timestamped sample for a pulsed shot.
#[derive(Clone, Copy, Debug)]
pub struct CampaignShotSample {
    pub t_s: f64,
    pub plasma: PulsedPlasmaTelemetry,
    pub bank: CapacitorBankTelemetry,
}

impl CampaignShotSample {
    /// Construct a validated sample.
    pub fn new(
        t_s: f64,
        plasma: PulsedPlasmaTelemetry,
        bank: CapacitorBankTelemetry,
    ) -> Result<Self, MultiShotCampaignError> {
        validate_non_negative("t_s", t_s)?;
        Ok(Self { t_s, plasma, bank })
    }
}

/// Input contract for one shot in a campaign.
#[derive(Clone, Debug)]
pub struct CampaignShotPlan {
    pub shot_id: String,
    pub samples: Vec<CampaignShotSample>,
    pub initial_bank_voltage_v: f64,
    pub initial_bank_current_a: f64,
    pub pulsed_mpc_admission_digest: Option<String>,
}

impl CampaignShotPlan {
    /// Construct a validated shot plan.
    pub fn new(
        shot_id: &str,
        samples: Vec<CampaignShotSample>,
        initial_bank_voltage_v: f64,
        initial_bank_current_a: f64,
    ) -> Result<Self, MultiShotCampaignError> {
        let id = shot_id.trim();
        if id.is_empty() {
            return Err(MultiShotCampaignError::EmptyShotId);
        }
        if samples.is_empty() {
            return Err(MultiShotCampaignError::EmptySamples);
        }
        validate_non_negative("initial_bank_voltage_v", initial_bank_voltage_v)?;
        validate_finite("initial_bank_current_a", initial_bank_current_a)?;
        Ok(Self {
            shot_id: id.to_string(),
            samples,
            initial_bank_voltage_v,
            initial_bank_current_a,
            pulsed_mpc_admission_digest: None,
        })
    }

    /// Attach digest-bound pulsed-MPC admission evidence to this shot plan.
    pub fn with_pulsed_mpc_admission_digest(
        mut self,
        digest: &str,
    ) -> Result<Self, MultiShotCampaignError> {
        self.pulsed_mpc_admission_digest = Some(validate_sha256_digest(
            "pulsed_mpc_admission_digest",
            digest,
        )?);
        Ok(self)
    }
}

/// Stable command log row.
#[derive(Clone, Debug)]
pub struct CampaignCommandLog {
    pub t_s: f64,
    pub state: String,
    pub action: String,
    pub reason: String,
    pub transition: bool,
    pub dwell_s: f64,
}

impl From<PulsedScenarioCommand> for CampaignCommandLog {
    fn from(command: PulsedScenarioCommand) -> Self {
        Self {
            t_s: command.t_s,
            state: command.state.as_str().to_string(),
            action: command.action.as_str().to_string(),
            reason: command.reason,
            transition: command.transition,
            dwell_s: command.dwell_s,
        }
    }
}

/// Stable phase-log row compatible with replay v1.1 shot metadata.
#[derive(Clone, Debug)]
pub struct CampaignPhaseLog {
    pub t_s: f64,
    pub state: String,
    pub reason: String,
}

/// Admission result for one shot.
#[derive(Clone, Debug)]
pub struct CampaignShotResult {
    pub shot_id: String,
    pub pulse_id: String,
    pub passed: bool,
    pub failure_reason: Option<String>,
    pub terminal_state: String,
    pub transition_states: Vec<String>,
    pub command_log: Vec<CampaignCommandLog>,
    pub shot_phase_log: Vec<CampaignPhaseLog>,
    pub capacitor_state_initial_j: f64,
    pub capacitor_state_final_j: f64,
    pub energy_recovered_j: f64,
    pub trigger_timestamp_ns: Option<u64>,
    pub pulsed_mpc_admission_digest: Option<String>,
}

/// Campaign report summary.
#[derive(Clone, Debug)]
pub struct MultiShotCampaignReport {
    pub schema_version: String,
    pub campaign_id: String,
    pub shot_count: usize,
    pub passed_count: usize,
    pub failed_count: usize,
    pub require_complete_lifecycle: bool,
    pub expected_transition_states: Vec<String>,
    pub pulsed_mpc_evidence_schema_version: String,
    pub pulsed_mpc_admission_digest_count: usize,
    pub shots: Vec<CampaignShotResult>,
    pub payload_sha256: String,
}

/// Multi-shot campaign orchestrator.
#[derive(Clone, Debug)]
pub struct MultiShotCampaignOrchestrator {
    campaign_id: String,
    scheduler_spec: PulsedScenarioSpec,
    bank_spec: CapacitorBankSpec,
    require_complete_lifecycle: bool,
}

impl MultiShotCampaignOrchestrator {
    /// Construct a validated orchestrator.
    pub fn new(
        campaign_id: &str,
        scheduler_spec: PulsedScenarioSpec,
        bank_spec: CapacitorBankSpec,
        require_complete_lifecycle: bool,
    ) -> Result<Self, MultiShotCampaignError> {
        let id = campaign_id.trim();
        if id.is_empty() {
            return Err(MultiShotCampaignError::EmptyCampaignId);
        }
        Ok(Self {
            campaign_id: id.to_string(),
            scheduler_spec,
            bank_spec,
            require_complete_lifecycle,
        })
    }

    /// Run all shots and return a digest-bound report.
    pub fn run(
        &self,
        shots: &[CampaignShotPlan],
    ) -> Result<MultiShotCampaignReport, MultiShotCampaignError> {
        if shots.is_empty() {
            return Err(MultiShotCampaignError::EmptyCampaign);
        }
        let mut seen = HashSet::with_capacity(shots.len());
        let mut results = Vec::with_capacity(shots.len());
        for shot in shots {
            if !seen.insert(shot.shot_id.clone()) {
                return Err(MultiShotCampaignError::DuplicateShotId(
                    shot.shot_id.clone(),
                ));
            }
            results.push(self.run_shot(shot)?);
        }
        let passed_count = results.iter().filter(|shot| shot.passed).count();
        let mut report = MultiShotCampaignReport {
            schema_version: MULTI_SHOT_CAMPAIGN_SCHEMA_VERSION.to_string(),
            campaign_id: self.campaign_id.clone(),
            shot_count: results.len(),
            passed_count,
            failed_count: results.len() - passed_count,
            require_complete_lifecycle: self.require_complete_lifecycle,
            expected_transition_states: EXPECTED_TRANSITION_STATES
                .iter()
                .map(|value| value.to_string())
                .collect(),
            pulsed_mpc_evidence_schema_version: PULSED_MPC_DECISION_EVIDENCE_SCHEMA_VERSION
                .to_string(),
            pulsed_mpc_admission_digest_count: results
                .iter()
                .filter(|shot| shot.pulsed_mpc_admission_digest.is_some())
                .count(),
            shots: results,
            payload_sha256: String::new(),
        };
        report.payload_sha256 = report_digest(&report);
        Ok(report)
    }

    fn run_shot(
        &self,
        shot: &CampaignShotPlan,
    ) -> Result<CampaignShotResult, MultiShotCampaignError> {
        let mut scheduler = PulsedScenarioScheduler::new(self.scheduler_spec);
        let bank = CapacitorBank::new(
            self.bank_spec,
            shot.initial_bank_voltage_v,
            shot.initial_bank_current_a,
        )
        .map_err(|err| MultiShotCampaignError::Bank(err.to_string()))?;
        let initial_energy = bank.state().energy_j();
        let mut last_energy = initial_energy;
        let mut min_energy = initial_energy;
        let mut command_log = Vec::with_capacity(shot.samples.len());
        let mut failure_reason: Option<String> = None;

        for sample in &shot.samples {
            last_energy = sample.bank.energy_j;
            min_energy = min_energy.min(last_energy);
            match scheduler.step(sample.t_s, sample.plasma, sample.bank) {
                Ok(command) => command_log.push(command.into()),
                Err(err) => {
                    failure_reason = Some(err.to_string());
                    break;
                }
            }
        }

        let shot_phase_log: Vec<CampaignPhaseLog> = scheduler
            .audit_log()
            .iter()
            .map(|record| CampaignPhaseLog {
                t_s: record.t_s,
                state: record.to_state.as_str().to_string(),
                reason: record.reason.clone(),
            })
            .collect();
        let transition_states: Vec<String> =
            shot_phase_log.iter().map(|row| row.state.clone()).collect();
        let terminal_state = scheduler.state.as_str().to_string();
        if failure_reason.is_none() {
            failure_reason = self.admission_failure_reason(&terminal_state, &transition_states);
        }
        let trigger_timestamp_ns = shot_phase_log
            .iter()
            .find(|row| row.state == "burn")
            .map(|row| (row.t_s * 1_000_000_000.0).round() as u64);
        Ok(CampaignShotResult {
            shot_id: shot.shot_id.clone(),
            pulse_id: pulse_id(&self.campaign_id, &shot.shot_id),
            passed: failure_reason.is_none(),
            failure_reason,
            terminal_state,
            transition_states,
            command_log,
            shot_phase_log,
            capacitor_state_initial_j: initial_energy,
            capacitor_state_final_j: last_energy,
            energy_recovered_j: (last_energy - min_energy).max(0.0),
            trigger_timestamp_ns,
            pulsed_mpc_admission_digest: shot.pulsed_mpc_admission_digest.clone(),
        })
    }

    fn admission_failure_reason(
        &self,
        terminal_state: &str,
        transition_states: &[String],
    ) -> Option<String> {
        if self.require_complete_lifecycle
            && transition_states
                != EXPECTED_TRANSITION_STATES
                    .iter()
                    .map(|value| value.to_string())
                    .collect::<Vec<String>>()
        {
            return Some("shot did not traverse the complete pulsed lifecycle".to_string());
        }
        if terminal_state != "idle" {
            return Some("shot did not return to idle".to_string());
        }
        None
    }
}

/// Campaign orchestration errors.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum MultiShotCampaignError {
    EmptyCampaignId,
    EmptyCampaign,
    EmptyShotId,
    EmptySamples,
    DuplicateShotId(String),
    NonFinite(&'static str),
    Negative(&'static str),
    InvalidDigest(&'static str),
    Bank(String),
    Scheduler(String),
}

impl From<PulsedScenarioError> for MultiShotCampaignError {
    fn from(value: PulsedScenarioError) -> Self {
        Self::Scheduler(value.to_string())
    }
}

impl Display for MultiShotCampaignError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::EmptyCampaignId => write!(f, "campaign_id must not be empty"),
            Self::EmptyCampaign => write!(f, "shots must not be empty"),
            Self::EmptyShotId => write!(f, "shot_id must not be empty"),
            Self::EmptySamples => write!(f, "samples must not be empty"),
            Self::DuplicateShotId(shot_id) => write!(f, "duplicate shot_id: {shot_id}"),
            Self::NonFinite(field) => write!(f, "{field} must be finite"),
            Self::Negative(field) => write!(f, "{field} must be non-negative"),
            Self::InvalidDigest(field) => {
                write!(f, "{field} must be a lowercase SHA-256 hex digest")
            }
            Self::Bank(reason) => write!(f, "{reason}"),
            Self::Scheduler(reason) => write!(f, "{reason}"),
        }
    }
}

impl Error for MultiShotCampaignError {}

fn validate_finite(name: &'static str, value: f64) -> Result<(), MultiShotCampaignError> {
    if !value.is_finite() {
        return Err(MultiShotCampaignError::NonFinite(name));
    }
    Ok(())
}

fn validate_non_negative(name: &'static str, value: f64) -> Result<(), MultiShotCampaignError> {
    validate_finite(name, value)?;
    if value < 0.0 {
        return Err(MultiShotCampaignError::Negative(name));
    }
    Ok(())
}

fn validate_sha256_digest(
    name: &'static str,
    value: &str,
) -> Result<String, MultiShotCampaignError> {
    let digest = value.trim();
    if digest.len() != 64
        || !digest
            .as_bytes()
            .iter()
            .all(|byte| byte.is_ascii_digit() || (*byte >= b'a' && *byte <= b'f'))
    {
        return Err(MultiShotCampaignError::InvalidDigest(name));
    }
    Ok(digest.to_string())
}

fn pulse_id(campaign_id: &str, shot_id: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(b"scpn-control.multi-shot-campaign.pulse-id.v1");
    hasher.update(campaign_id.as_bytes());
    hasher.update([0]);
    hasher.update(shot_id.as_bytes());
    let digest = hasher.finalize();
    let mut bytes = [0_u8; 16];
    bytes.copy_from_slice(&digest[..16]);
    bytes[6] = (bytes[6] & 0x0f) | 0x50;
    bytes[8] = (bytes[8] & 0x3f) | 0x80;
    format!(
        "{:02x}{:02x}{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
        bytes[0],
        bytes[1],
        bytes[2],
        bytes[3],
        bytes[4],
        bytes[5],
        bytes[6],
        bytes[7],
        bytes[8],
        bytes[9],
        bytes[10],
        bytes[11],
        bytes[12],
        bytes[13],
        bytes[14],
        bytes[15]
    )
}

fn report_digest(report: &MultiShotCampaignReport) -> String {
    let mut hasher = Sha256::new();
    hasher.update(report.schema_version.as_bytes());
    hasher.update(report.campaign_id.as_bytes());
    hasher.update(report.shot_count.to_string().as_bytes());
    hasher.update(report.passed_count.to_string().as_bytes());
    hasher.update(report.failed_count.to_string().as_bytes());
    hasher.update(report.pulsed_mpc_evidence_schema_version.as_bytes());
    hasher.update(
        report
            .pulsed_mpc_admission_digest_count
            .to_string()
            .as_bytes(),
    );
    for shot in &report.shots {
        hasher.update(shot.shot_id.as_bytes());
        hasher.update(shot.pulse_id.as_bytes());
        hasher.update(shot.passed.to_string().as_bytes());
        hasher.update(shot.terminal_state.as_bytes());
        if let Some(digest) = &shot.pulsed_mpc_admission_digest {
            hasher.update(digest.as_bytes());
        }
        for state in &shot.transition_states {
            hasher.update(state.as_bytes());
        }
    }
    format!("{:x}", hasher.finalize())
}
